### PR TITLE
[onert] Introduce ir::train::TrainingInfo

### PR DESCRIPTION
**Pull request by zetwhite**
This PR introduces ir::train::TrainingInfo into core.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

related issue : https://github.com/Samsung/ONE/issues/11692 
draft : https://github.com/Samsung/ONE/pull/12152

### CODE REVIEW

**Comment by jyoungyun** [comment]
(suggestion)
How about separating this PR into two PRs, one is to define `ir::train::TrainingInfo` and the other is to add  `TrainingInfo` to `nnfw_session`. And is changing `compile::train::TrainingInfo` to `ir::train::TrainingInfo` a big change? If you don't mind, I suggest updating the codes related to `compile::train::TrainingInfo`  at once in one pr.

File: runtime/onert/core/src/ir/train/TrainingInfo.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    loss_info.reduction_type = ir::train::LossReductionType::SumOverBatchSize;
+  }
+  info->setLossInfo(loss_info);
+
+  onert::ir::train::OptimizerInfo opt_info;
+  {
+    opt_info.optim_code = ir::train::OptimizerCode::SGD;
+    opt_info.learning_rate = 0.001f;
+  }
+  info->setOptimizerInfo(opt_info);
```
Comment (jyoungyun) [inline runtime/onert/core/src/ir/train/TrainingInfo.cc:44]: Should we initialize these variables? I think there is no need to redefine it here because there is already an initial value.

File: runtime/onert/core/include/ir/train/TrainingInfo.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+{
+namespace train
+{
+
+class TrainingInfo final
+{
+public:
+  TrainingInfo() : _loss_info(), _optimizer_info(), _batch_size(0), _epoch(0) {}
+  // named constructor : create a default valid TrainingInfo
+  static std::unique_ptr<TrainingInfo> createDefaultInfo();
```
Comment (zetwhite) [inline runtime/onert/core/include/ir/train/TrainingInfo.h:39]: Q: Why did you use the named constructor?  

A: I think it is confusing that some `ir` default-constructors make an invalid one, but sometimes not.  
So, I prefer to give a named constructor, to show the caller that this `ir` is filled with default 'valid' value.

File: runtime/onert/core/include/ir/train/TrainingInfo.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  uint32_t batchSize() const { return _batch_size; }
+  uint32_t epoch() const { return _epoch; }
+
+  // setter
+  void setBatchSize(const uint32_t batch_size) { _batch_size = batch_size; }
+  void setLossInfo(const LossInfo &loss_info) { _loss_info = loss_info; }
+  void setOptimizerInfo(const OptimizerInfo &optimizer_info) { _optimizer_info = optimizer_info; }
+  void setEpoch(uint32_t epoch) { _epoch = epoch; }
+
+  bool isValid() const;
```
Comment (zetwhite) [inline runtime/onert/core/include/ir/train/TrainingInfo.h:55]: I'd like to check the TrainingInfo validation, using this method in `train_prepare`. 
in draft, `isValid()` is used here - https://github.com/Samsung/ONE/pull/12152/files#r1432281266 

We can check the "combination of each field" is fine with this method.

File: runtime/onert/api/src/nnfw_api_internal.cc [hunk 4]
```diff
@@ -400,6 +406,8 @@ NNFW_STATUS nnfw_session::load_model_from_nnpackage(const char *package_dir)
       model->bindKernelBuilder(_kernel_registry->getBuilder());
       _nnpkg->push(onert::ir::ModelIndex{i}, std::move(model));
       _coptions.push_back(onert::compiler::CompilerOptions::fromGlobalConfig());
+      // TODO load TrainingInfo from model, using TraininfoLoader
+      _train_info = onert::ir::train::TrainingInfo::createDefaultInfo();
```
Comment (zetwhite) [inline runtime/onert/api/src/nnfw_api_internal.cc:410]: I have a plan about TODO comment like this. 
A draft code related with this TODO - https://github.com/Samsung/ONE/pull/12152/files#r1426184845

**Comment by zetwhite** [comment]
> How about separating this PR into two PRs, one is to define `ir::train::TrainingInfo` and the other is to add `TrainingInfo` to `nnfw_session`. 

Thank you for the suggestion! No problem. I'll separate it into two :) 

> And is changing `compile::train::TrainingInfo` to `ir::train::TrainingInfo` a big change? If you don't mind, I suggest updating the codes related to `compile::train::TrainingInfo` at once in one pr.

Originally, I'd like to complete loader-things first. and replace compile::TrainingInfo with `ir` later.
Because 
   - replacing is not really necessary for the loader.
   - `compile::train::TrainingInfo` is used in Compiler, Executor and TrainableOperationConverter,
      Since I'm not familiar with these, I'd like to replace it carefully. 

So, It is fine, I'd like to replace `compiler::TraininnInfo` with `ir::TrainingInfo` later. 
But if you think it is better to do now, I'd like to follow. So, please share your opinion. Thank you :

File: runtime/onert/core/src/ir/train/TrainingInfo.cc (hunk 1 above)
Comment (zetwhite) [reply runtime/onert/core/src/ir/train/TrainingInfo.cc:44]: I append my idea about the named constructor here - https://github.com/Samsung/ONE/pull/12340/files/21d8ef90f0277b3f9df5f822d26cfd8adf56372a#r1432279409. 

The initial constructor has initially "invalid" value by default. 
The named constructor returns a default "valid" info, to be used when training_info isn't given by a model or user.

File: runtime/onert/core/include/ir/train/TrainingInfo.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#include "ir/train/LossInfo.h"
+
+namespace onert
+{
+namespace ir
+{
+namespace train
+{
+
+class TrainingInfo final
```
Comment (zetwhite) [inline runtime/onert/core/include/ir/train/TrainingInfo.h:34]: (plan) This TrainingInfo will be saved in nnfw_session : https://github.com/Samsung/ONE/pull/12342/commits/8fa3ffaec360916b60eba426e533da61952fa9e0

**Comment by jyoungyun** [comment]
> Thank you for the suggestion! No problem. I'll separate it into two :) 

Thank you :)

> Originally, I planned to complete loader-things first. and replace `compile::TrainingInfo` with `ir` later.
> Because 
>    - replacing is not really necessary for the loader.
>    - `compile::train::TrainingInfo` is used in Compiler, Executor and TrainableOperationConverter,
>       Since I'm not familiar with these, I'd like to replace it carefully. 
> 
> So, It is fine, I'd like to replace `compiler::TraininnInfo` with `ir::TrainingInfo` later. 

I respect your opinion. It may take more time to replace `compiler::TrainingInfo` with `ir::TrainingInfo`. So we continue this change as it is and apply for replacement later. :)

File: runtime/onert/core/src/ir/train/TrainingInfo.cc (hunk 1 above)
Comment (zetwhite) [reply runtime/onert/core/src/ir/train/TrainingInfo.cc]: I removed the default valid training info after an offline discussion.

File: runtime/onert/core/include/ir/train/TrainingInfo.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_IR_TRAIN_TRAINING_INFO_H__
+#define __ONERT_IR_TRAIN_TRAINING_INFO_H__
+
+#include <memory>
```
Comment (jyoungyun) [inline runtime/onert/core/include/ir/train/TrainingInfo.h:21]:
```suggestion
```

**Review by jyoungyun — APPROVED** [review]
LGTM Thank you!

**Comment by zetwhite** [comment]
> LGTM Thank you!

updated once more, because of code formatting issue :sweat_smile:

**Review by glistening — APPROVED** [review]
LGTM

File: runtime/onert/core/src/ir/train/TrainingInfo.cc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+bool TrainingInfo::isValid() const
+{
+  if (_batch_size == 0)
+    return false;
+
+  if (_optimizer_info.optim_code == onert::ir::train::OptimizerCode::Invalid)
+    return false;
+
+  if (_optimizer_info.learning_rate == 0.0f)
```
Comment (ragmani) [inline runtime/onert/core/src/ir/train/TrainingInfo.cc:34]: If learning rate is a negative value, outputs will move away from the minimum. In other words, outputs will move to reverse what gradient descent does, and nn model won't be trained.
So, it would be better to change the condition if this function is for checking valid info of `TrainingInfo`, not default values.
```suggestion
  if (_optimizer_info.learning_rate <= 0.0f)
```

**Review by ragmani — APPROVED** [review]

**Review by jyoungyun — APPROVED** [review]
LGTM

File: runtime/onert/core/src/ir/train/TrainingInfo.cc (hunk 7 above)
Comment (glistening) [reply runtime/onert/core/src/ir/train/TrainingInfo.cc:34]: @zetwhite or @ragmani Please update this in another PR.
